src/Utilities/TrackedDependencies/CanonicalTrackedInputFiles.cs

```diff
@@ -1070,8 +1073,18 @@ private void RemoveDependenciesFromEntryIfMissing(string rootingMarker)
                 {
                     if (keyIndex++ > 0)
                     {
-                        // If we are ignoring missing files, then only record those that exist
-                        if (FileUtilities.FileExistsNoThrow(file))
+                        // Record whether or not each file exists and cache it.
+                        // We do this to save time (On^2), at the expense of data O(n).
+                        bool inFileCache = fileCache.TryGetValue(file, out bool fileExists);
+
+                        // Have we cached the file yet? If not, cache whether or not it exists.
+                        if(!inFileCache)
+                        {
+                            fileCache.Add(file, FileUtilities.FileExistsNoThrow(file));
```

Comment:
`fileExists` is not set on cache miss. It's going to have the default value of `false` even if the file exists.

Reply:
Good catch, fixed.